### ISSUE TITLE
Possible bug with `requires_grad` attribute being changed following operations

### ISSUE THREAD

**Issue by jatkinson1000** [Cambridge-ICCS/FTorch#370]
I have observed a case where the `requires_grad` attribute for a tensor is flipped from `.false.` to `.true.` as a result of assigning it based on a calculation involving other tensors.

I have created a MWE to showcase this and will open a PR to demonstrate.

The key lines are that:

```fortran
    output_vec = input_vec * scaling_tensor
    grad_reqd = loss%requires_grad()
    write(*,*) "loss%requires_grad = ", grad_reqd

    loss = ((output_vec - target_vec) ** 2)
    grad_reqd = loss%requires_grad()
    write(*,*) "loss%requires_grad = ", grad_reqd
```
cause `requires_grad` for `loss` to go from `.false.` to `.true.`.

The only tensor in the entire example that has `requires_grad=.true.` set is `scaling tensor`.

This is an issue because it prevents values being assigned to `loss` in subsequent calculations.

**Comment by jatkinson1000** [comment]
I have added a MWE to the testing workflow in #371 that showcases this.
